### PR TITLE
perf: ioring background processing fixes

### PR DESCRIPTION
**Pull request by sulakshm**
Signed-off-by: Lakshmi Narasimhan Sundararajan <lns@portworx.com>

**What this PR does / why we need it**:
This is my in a series of change to address perf gap between master and 2.8.
This focusses only on uring path changes.

1/ enables background processing of IOs submitted to uring. Currently they are run in the context of the 
submit user thread.

2/ below results confirm perf is better than master in all profiles.
3/ vs 2.8 still some work to do.

This goes with https://github.com/portworx/porx/pull/7913

**Which issue(s) this PR fixes** (optional)
Closes #
https://portworx.atlassian.net/browse/PWX-20970
**Special notes for your reviewer**:

                 | 2.8.0 default | master default | only uring changes |     
1vol          | KIOPS           | KIOPS               | KIOPS                      | % improvement vs 2.8 | % improvement vs master
seqwrite   | 172               | 115                   | 116                          | -32.55813953               | 0.869565217
seqread    | 246               | 246                   |  296                         | 20.32520325                | 20.32520325
randwrite | 87                 | 68                     | 71                            | -18.3908046                 | 4.411764706
randread  | 137               | 93                     | 270                          | 97.08029197                 | 190.3225806

40vol        
seqwrite   | 234               | 312                  | 357                           | 52.56410256                  | 14.42307692
seqread    | 420               | 184                 | 185                            | -55.95238095                 | 0.543478261
randwrite | 95                 | 81                    | 83                             | -12.63157895                 | 2.469135802
randread  | 62                | 73                     | 183                           | 195.1612903                   | 150.6849315

### CODE REVIEW

File: fuse_i.h [hunk 1]
```diff
@@ -141,7 +141,7 @@ struct ____cacheline_aligned fuse_queue_reader {
 	uint32_t read;          /** read index updated by reader */
 	uint32_t write;		/** write index updated by writer */
 	uint32_t pad_0;
-	uint32_t pad;
+	atomic_t need_wake_up;
```
Comment (sulakshm) [inline fuse_i.h:144]: it is the reader that needs wakeup. correcting semantics of definition.

File: io.c [hunk 2]
```diff
@@ -1459,18 +1460,13 @@ static int io_req_defer(struct io_ring_ctx *ctx, struct io_kiocb *req,
 	if (!io_sequence_defer(ctx, req) && list_empty(&ctx->defer_list))
 		return 0;
 
-	sqe_copy = kmalloc(sizeof(*sqe_copy), GFP_KERNEL);
-	if (!sqe_copy)
-		return -EAGAIN;
-
+	sqe_copy = &req->cached_sqe;
```
Comment (sulakshm) [inline io.c:1463]: every req now has a cached copy of the original req. This saves later memory allocation, all memory needs are allocated and initialized upfront.

File: io.c [hunk 3]
```diff
@@ -1793,31 +1788,26 @@ static int io_submit_sqe(struct io_ring_ctx *ctx, struct sqe_submit *s,
...
-				queue_work(ctx->sqo_wq, &req->work);
-			}
-
-			/*
-			 * Queued up for async execution, worker will release
-			 * submit reference when the iocb is actually
-			 * submitted.
-			 */
-			return 0;
+		struct io_uring_sqe *sqe_copy = &req->cached_sqe;
```
Comment (sulakshm) [inline io.c:1791]: this code path gets simplified using the cached copy instead of memory allocation later.

File: io.c [hunk 4]
```diff
@@ -1936,16 +1926,57 @@ static int io_submit_sqes(struct io_ring_ctx *ctx, struct sqe_submit *sqes,
 	return submitted;
 }
 
+static mm_segment_t io_sq_setup_user_mm(struct mm_struct *mm, bool *mm_fault)
```
Comment (sulakshm) [inline io.c:1929]: wrappers to handle userspace locking/unlocking.

File: io.c [hunk 5]
```diff
@@ -2232,6 +2229,8 @@ static int io_sqe_files_register(struct io_ring_ctx *ctx, void __user *arg,
 static int io_sq_offload_start(struct io_ring_ctx *ctx, struct io_uring_params *p)
 {
 	int ret;
+	int i;
+	uint32_t max_threads = min(p->sqo_threads, NSLAVES);
```
Comment (sulakshm) [inline io.c:2233]: max threads assigned is now part of params.

File: io.c [hunk 6]
```diff
@@ -2725,6 +2727,32 @@ static long io_ring_ioctl_init(struct io_ring_ctx *ctx, unsigned long arg)
 	return 0;
 }
 
+static long io_run_cmd(struct io_ring_ctx *ctx, unsigned long arg)
```
Comment (sulakshm) [inline io.c:2730]: new ioctl exported to submit and process reqs immediately.

**Review by sulakshm — COMMENTED** [review]
added comments

**Review by maxkozlovsky — COMMENTED** [review]
There are way too many things going on here. You should submit PRs which do one logically complete thing. It is hard to give meaningful feedback when there are 5 different unrelated design decisions in a single PR.

**Pull request closed without merging**